### ISSUE TITLE
[BUG]  Memory Allocation with C++

### ISSUE THREAD

**Issue by boureau93** [cython/cython#5936]
### Describe the bug

I've been working with Cython for a while and recently started working with C++ and wrapping C++ with Cython. I'm building a class that holds some pointers that are then dynamically allocated inside the constructor and have had some problems regarding memory management (lots of double free, corruption, invalid pointer and segfault errors).  After some experimenting I've found some minimal code to reproduce this bug.  Below you can see my layer.hpp, layer.cpp, mlodecl.pxd, mlo.pyx and setup.py files that produced his error.

### Code to reproduce the behaviour:

```layer.hpp

#ifndef LAYER_H
#define LAYER_H

class layer{

public:
    int nodes;
    float* h; //auxiliary field
    float* mean; //mean of a layer

    ~layer();
    layer();
    layer(int);

    bool isNull();
};

#endif
```

```layer.cpp

#include "layer.hpp"

layer::layer(){}

layer::layer(int nodes){
    this->nodes = nodes;
    this->h = new float[nodes];
}

layer::~layer(){
    if (this->h!=nullptr)
        delete[] this->h;
}

bool layer::isNull(){
    if (this->h==nullptr)
        return true;
    return false;
}
```

```mlo.pyx

cimport mlodecl as ml
cimport cython


cdef class Layer:

    cdef ml.layer L

    def __init__(self, int nodes):
        self.L = ml.layer(nodes)
        if self.L.isNull():
            raise Exception("Failed to allocate memory")

    @property
    def numNodes(self):
        return self.L.nodes
```

```mlodecl.pxd

cdef extern from "cfiles/layer.hpp":

    cdef cppclass layer:
        int nodes
        float* h

        layer()
        layer(int)

        bint isNull()
```
```setup.py

from setuptools import setup, Extension
from Cython.Build import cythonize
import numpy as np
import os

sourcefiles = ["mlo.pyx","cfiles/layer.cpp"]

os.environ["CXX"] = "g++"

ext = Extension('miolo', sourcefiles, language="c++",
        extra_compile_args=['-fopenmp','-lm'],
        extra_link_args=['-fopenmp'],
        library_dirs=["/usr/lib/gcc/x86_64-linux-gnu/12"],
        libraries=["gomp"],
        include_dirs=[np.get_include()])

setup(ext_modules=cythonize(ext))
```

### Expected behaviour

Basically, allocation and deallocation without errors can only be done if my Layer C++ attribute is made a pointer and then dynamically allocated using new, which also requires a __dealloc__ method. But when I do it inside only an __init__ method (stack allocated object), it seems memory allocation is fine, but when I close Python I get a double free or corruption (top).  If I change the cdef ml.Layer to cdef ml.Layer* and add the appropriate dealloc method, the problem vanishes.

### OS

Debian 12

### Python version

3.11.2

### Cython version

3.0.7

### Additional context

I was also having problems with the __init__ method previously (double free when initializing in Python). This disappeared when I removed the except+ in the declaration of constructors inside my pxd file.

**Comment by da-woods** [comment]
What's happening here is:

1. An instance of `Layer` is default constructed inside your cdef class.
2. You create a temporary instance of `Layer` and assign that (`operator=`) to the one stored in your cdef class.
3. The temporary instance of layer is destructed, freeing the memory.
4. Your cdef class is destructed, freeing the memory again.

This is a bug in your code, not a bug in Cython. Read https://en.cppreference.com/w/cpp/language/rule_of_three.

Support questions are best directed to the Cython-user's mailing list in future (I appreciate you probably thought this was a Cython bug though)

**Comment by boureau93** [comment]
Really insightful comment of yours! So glad this is not a real bug. I've tried so hard and thought this could only be due to a malfunction. Thank you very much.
